File: Parcel-Delivery-Tracking-System-main/frontend/algorithms.js
/**
 * ALGORITHMS MODULE
 * Implements: Linear Search, Binary Search, Bubble Sort, Selection Sort, Merge Sort
 */

// ============================================
// LINEAR SEARCH
// Purpose: Find parcel by searching every parcel one by one
// Time Complexity: O(n)
// Space Complexity: O(1)
// ============================================
function linearSearch(parcels, fieldName, searchValue) {
  // Iterate through every parcel
  for (let i = 0; i < parcels.length; i++) {
    // Check if current parcel's field matches search value
    if (parcels[i][fieldName] === searchValue) {
      return parcels[i]; // Found - return immediately
    }
  }
  return null; // Not found
}

// ============================================
// BINARY SEARCH
// Purpose: Fast search on sorted parcels by tracking number
// Time Complexity: O(log n)
// Space Complexity: O(1)
// ============================================
function binarySearch(sortedParcels, trackingNumber) {
  let left = 0;
  let right = sortedParcels.length - 1;

  // Continue while search space exists
  while (left <= right) {
    let mid = Math.floor((left + right) / 2);
    let midValue = sortedParcels[mid].trackingNumber;

    // Check middle element
    if (midValue === trackingNumber) {
      return sortedParcels[mid]; // Found
    } else if (midValue < trackingNumber) {
      left = mid + 1; // Search right half
    } else {
      right = mid - 1; // Search left half
    }
  }
  return null; // Not found
}

// ============================================
// BUBBLE SORT
// Purpose: Simple sorting for educational demonstration
// Time Complexity: O(n²)
// Space Complexity: O(1)
// ============================================
function bubbleSort(parcels, fieldName) {
  let arr = [...parcels]; // Copy array
  let n = arr.length;

  // Outer loop: repeat n times
  for (let i = 0; i < n - 1; i++) {
    // Inner loop: compare adjacent elements
    for (let j = 0; j < n - i - 1; j++) {
      // Swap if current > next
      if (arr[j][fieldName] > arr[j + 1][fieldName]) {
        let temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
      }
    }
  }
  return arr;
}

// ============================================
// SELECTION SORT
// Purpose: Sorting for demonstration
// Time Complexity: O(n²)
// Space Complexity: O(1)
// ============================================
function selectionSort(parcels, fieldName) {
  let arr = [...parcels]; // Copy array
  let n = arr.length;

  // Outer loop: select minimum for each position
  for (let i = 0; i < n - 1; i++) {
    let minIndex = i;

    // Inner loop: find minimum element
    for (let j = i + 1; j < n; j++) {
      if (arr[j][fieldName] < arr[minIndex][fieldName]) {
        minIndex = j;
      }
    }

    // Swap minimum with current position
    if (minIndex !== i) {
      let temp = arr[i];
      arr[i] = arr[minIndex];
      arr[minIndex] = temp;
    }
  }
  return arr;
}

// ============================================
// MERGE SORT
// Purpose: Efficient sorting for larger datasets
// Time Complexity: O(n log n)
// Space Complexity: O(n)
// ============================================
function mergeSort(parcels, fieldName) {
  // Base case: array of 1 or 0 elements is sorted
  if (parcels.length <= 1) {
    return parcels;
  }

  // Divide: split array in half
  let mid = Math.floor(parcels.length / 2);
  let left = parcels.slice(0, mid);
  let right = parcels.slice(mid);

  // Conquer: recursively sort both halves
  left = mergeSort(left, fieldName);
  right = mergeSort(right, fieldName);

  // Combine: merge sorted halves
  return merge(left, right, fieldName);
}

// Helper function to merge two sorted arrays
function merge(left, right, fieldName) {
  let result = [];
  let i = 0, j = 0;

  // Compare elements from left and right, add smaller to result
  while (i < left.length && j < right.length) {
    if (left[i][fieldName] <= right[j][fieldName]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  // Add remaining elements
  while (i < left.length) {
    result.push(left[i]);
    i++;
  }
  while (j < right.length) {
    result.push(right[j]);
    j++;
  }

  return result;
}

File: backend/algorithms.py
"""
ALGORITHMS MODULE
==================
Implements: Linear Search, Binary Search, Bubble Sort, Selection Sort,
Merge Sort, and Dijkstra's Shortest Path.

All functions take/return plain dicts (parcels) or a Graph (see
data_structures.py) so they can be called directly from the Flask
routes in app.py and their result shown in the UI.
"""

import heapq


# ============================================================
# LINEAR SEARCH
# Purpose : find a parcel by scanning every record - works on any
#           field, sorted or not
# Time    : O(n)
# Space   : O(1)
# ============================================================
def linear_search(parcels, field, value):
    for parcel in parcels:
        if str(parcel.get(field, '')).lower() == str(value).lower():
            return parcel
    return None


# ============================================================
# BINARY SEARCH
# Purpose : fast lookup once parcels are sorted by the search field
# Time    : O(log n)
# Space   : O(1)
# Note    : caller must pass a list already sorted (ascending) on
#           `field` - see merge_sort() below.
# ============================================================
def binary_search(sorted_parcels, field, value):
    left, right = 0, len(sorted_parcels) - 1
    target = str(value).lower()

    while left <= right:
        mid = (left + right) // 2
        mid_value = str(sorted_parcels[mid].get(field, '')).lower()

        if mid_value == target:
            return sorted_parcels[mid]
        elif mid_value < target:
            left = mid + 1
        else:
            right = mid - 1
    return None


# ============================================================
# BUBBLE SORT
# Purpose : simple comparison sort, kept for the algorithm defense
# Time    : O(n^2)
# Space   : O(1) extra (sorts a copy, in place)
# ============================================================
def bubble_sort(parcels, field, reverse=False):
    arr = list(parcels)
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            should_swap = arr[j][field] > arr[j + 1][field]
            if reverse:
                should_swap = arr[j][field] < arr[j + 1][field]
            if should_swap:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr


# ============================================================
# SELECTION SORT
# Purpose : second simple comparison sort, for side-by-side comparison
# Time    : O(n^2)
# Space   : O(1) extra
# ============================================================
def selection_sort(parcels, field, reverse=False):
    arr = list(parcels)
    n = len(arr)
    for i in range(n - 1):
        target_index = i
        for j in range(i + 1, n):
            better = arr[j][field] < arr[target_index][field]
            if reverse:
                better = arr[j][field] > arr[target_index][field]
            if better:
                target_index = j
        if target_index != i:
            arr[i], arr[target_index] = arr[target_index], arr[i]
    return arr


# ============================================================
# MERGE SORT
# Purpose : efficient divide-and-conquer sort, used to prepare the
#           list for binary_search() above
# Time    : O(n log n)
# Space   : O(n)
# ============================================================
def merge_sort(parcels, field, reverse=False):
    if len(parcels) <= 1:
        return list(parcels)

    mid = len(parcels) // 2
    left = merge_sort(parcels[:mid], field, reverse)
    right = merge_sort(parcels[mid:], field, reverse)
    return _merge(left, right, field, reverse)


def _merge(left, right, field, reverse):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        take_left = left[i][field] <= right[j][field]
        if reverse:
            take_left = left[i][field] >= right[j][field]
        if take_left:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


# ============================================================
# DIJKSTRA'S SHORTEST PATH
# Purpose : find the shortest road distance between two towns in the
#           delivery network (Graph from data_structures.py); the
#           result feeds the cost calculator's distance_km figure
# Time    : O((V + E) log V) with a binary heap as the priority queue
# Space   : O(V)
# ============================================================
def dijkstra(graph, start, end):
    if start not in graph.adjacency or end not in graph.adjacency:
        return None, float('inf')

    distances = {town: float('inf') for town in graph.adjacency}
    previous = {town: None for town in graph.adjacency}
    distances[start] = 0

    visited = set()
    heap = [(0, start)]

    while heap:
        current_distance, current_town = heapq.heappop(heap)
        if current_town in visited:
            continue
        visited.add(current_town)

        if current_town == end:
            break

        for neighbour, weight in graph.neighbours(current_town):
            new_distance = current_distance + weight
            if new_distance < distances[neighbour]:
                distances[neighbour] = new_distance
                previous[neighbour] = current_town
                heapq.heappush(heap, (new_distance, neighbour))

    if distances[end] == float('inf'):
        return None, float('inf')

    # Walk the `previous` chain backwards to rebuild the path
    path = []
    node = end
    while node is not None:
        path.append(node)
        node = previous[node]
    path.reverse()

    return path, distances[end]
